### PR TITLE
[Agent] Supports configure L7 log blacklist

### PR DESCRIPTION
**Pull request by TomatoMr**
### This PR is for:
#5916 
- Agent
- Server
### Supports configure L7 log blacklist
- Configs: 
  - Add L7LogBlacklist: map[string][]*L7LogBlacklist and UnconcernedDnsNxdomainResponseSuffixes: []string.
- L7ProtocolInfo:
  - Add a field `is_on_blacklist` to all `L7ProtocolInfo` to determine whether this L7 log is on the blacklist.
  - Add a field `last_is_on_blacklist` to all `L7ProtocolParserInterface` to determine if the previous L7 log was on the blacklist. If so, there is a high probability that the current L7 log should also not be counted for perf stats data.
  - After parsing the payload of all L7 logs, decide on the perf stats based on whether they are on the blacklist.
  - In the `parser.rs` file, check `l7_info.is_on_blacklist()` before sending `Box<MetaAppProto>`, rather than checking whether to send an L7 log after each `parse_payload`. This allows for the aggregation of L7 logs before making a collective judgment.
#### Checklist
- [ ] Added unit test.
#### Backport to branches
- main
- v6.5

### CODE REVIEW

File: agent/src/config/handler.rs [hunk 1]
```diff
@@ -692,6 +693,171 @@ impl From<&HttpEndpointExtraction> for HttpEndpointTrie {
...
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct BlacklistTrieNode {
+    children: HashMap<char, Box<BlacklistTrieNode>>,
+    operator: Option<Operator>,
+}
+
+impl BlacklistTrieNode {
+    pub fn is_on_blacklist(&self, input: &str) -> bool {
+        if input.is_empty() {
```
Comment (rvql) [inline agent/src/config/handler.rs:710]: op 空可以直接返回

File: agent/src/config/handler.rs [hunk 2]
```diff
@@ -692,6 +693,171 @@ impl From<&HttpEndpointExtraction> for HttpEndpointTrie {
...
+    const ENDPOINT: &'static str = "endpoint";
+    const REQUEST_TYPE: &'static str = "request_type";
+    const REQUEST_DOMAIN: &'static str = "request_domain";
+    const REQUEST_RESOURCE: &'static str = "request_resource";
+
+    // Currently, the following matching operations are supported:
+    const EQUAL: &'static str = "equal";
+    const PREFIX: &'static str = "prefix";
+
+    pub fn new_tries(
```
Comment (rvql) [inline agent/src/config/handler.rs:754]:
```suggestion
    pub fn new(
```

Comment (rvql) [reply agent/src/config/handler.rs:754]: HashMap 的逻辑跟 BlacklistTrie 没有关系，不应该揉在一起

File: agent/src/config/handler.rs [hunk 3]
```diff
@@ -692,6 +693,171 @@ impl From<&HttpEndpointExtraction> for HttpEndpointTrie {
...
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum Operator {
+    Equal,
+    Prefix,
+}
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct BlacklistTrieNode {
+    children: HashMap<char, Box<BlacklistTrieNode>>,
+    operator: Option<Operator>,
```
Comment (rvql) [inline agent/src/config/handler.rs:705]: 这个为啥要定义成 `Option`？

File: agent/src/config/handler.rs [hunk 4]
```diff
@@ -692,6 +693,171 @@ impl From<&HttpEndpointExtraction> for HttpEndpointTrie {
...
+    }
+
+    pub fn is_unconcerned(&self, input: &str) -> bool {
+        if input.is_empty() {
+            return false;
+        }
+        let mut node = &self.root;
+        let mut matched = false;
+        for c in input.chars().rev() {
+            // the reversal is because what is matched is the suffix of the domain name
```
Comment (rvql) [inline agent/src/config/handler.rs:838]: 注释位置不对

File: agent/src/config/handler.rs [hunk 5]
```diff
@@ -692,6 +693,171 @@ impl From<&HttpEndpointExtraction> for HttpEndpointTrie {
...
+            if let Some(child) = node.children.get(&c) {
+                matched = child.unconcerned; // if the child is a leaf, then matched
```
Comment (rvql) [inline agent/src/config/handler.rs:840]: 直接 return

File: agent/src/config/handler.rs [hunk 6]
```diff
@@ -692,6 +693,171 @@ impl From<&HttpEndpointExtraction> for HttpEndpointTrie {
...
+                if matched {
+                    break;
+                }
+                node = child.as_ref();
+            } else {
```
Comment (rvql) [inline agent/src/config/handler.rs:845]: 俩分支都处理用 match

File: agent/src/flow_generator/protocol_logs/dns.rs [hunk 7]
```diff
@@ -144,6 +154,18 @@ impl DnsInfo {
...
+
+    fn set_is_on_blacklist(&mut self, config: &LogParserConfig) {
+        if let Some(t) = config.l7_log_blacklist_trie.get(&L7Protocol::DNS) {
+            if t.request_resource.is_on_blacklist(&self.query_name)
+                || t.request_type.is_on_blacklist(self.get_domain_str())
+                || t.request_domain.is_on_blacklist(&self.query_name)
+                || t.endpoint.is_on_blacklist(&self.query_name)
+            {
+                self.is_on_blacklist = true;
+            }
```
Comment (rvql) [inline agent/src/flow_generator/protocol_logs/dns.rs:166]:
```suggestion
            self.is_on_blacklist = t.request_resource.is_on_blacklist(&self.query_name)
                || t.request_type.is_on_blacklist(self.get_domain_str())
                || t.request_domain.is_on_blacklist(&self.query_name)
                || t.endpoint.is_on_blacklist(&self.query_name);
```

Comment (rvql) [reply agent/src/flow_generator/protocol_logs/dns.rs:166]: 其他协议的也一样

File: agent/src/flow_generator/protocol_logs/mq/amqp.rs [hunk 8]
```diff
@@ -780,6 +810,29 @@ impl L7ProtocolInfoInterface for AmqpInfo {
     fn get_request_domain(&self) -> String {
         self.vhost.clone().unwrap_or_default()
     }
+
+    fn get_endpoint(&self) -> Option<String> {
+        let type1_len = self.exchange.as_ref().map_or(0, |r| r.len())
```
Comment (rvql) [inline agent/src/flow_generator/protocol_logs/mq/amqp.rs:815]: 这个变量名是啥含义？

File: agent/src/flow_generator/protocol_logs/mq/amqp.rs [hunk 9]
```diff
@@ -780,6 +810,29 @@ impl L7ProtocolInfoInterface for AmqpInfo {
...
+            + self.routing_key.as_ref().map_or(0, |r| r.len());
+        let endpoint = if type1_len > 0 {
+            format!(
+                "{}.{}",
+                self.exchange.clone().unwrap_or_default(),
```
Comment (rvql) [inline agent/src/flow_generator/protocol_logs/mq/amqp.rs:820]: clone 操作多余 copy 了字符串

File: agent/src/flow_generator/protocol_logs/mq/amqp.rs [hunk 10]
```diff
@@ -780,6 +810,29 @@ impl L7ProtocolInfoInterface for AmqpInfo {
...
+                self.routing_key.clone().unwrap_or_default()
+            )
+        } else {
+            self.queue.clone().unwrap_or_default()
+        };
+        if endpoint.is_empty() {
+            None
+        } else {
+            Some(endpoint)
+        }
```
Comment (rvql) [inline agent/src/flow_generator/protocol_logs/mq/amqp.rs:830]:
```suggestion
        if type1_len == 0 {
            return self.queue.clone().unwrap_or_default();
        }
        Some(format!(
                "{}.{}",
                self.exchange.clone().unwrap_or_default(),
                self.routing_key.clone().unwrap_or_default()
        ))
```

File: agent/src/flow_generator/protocol_logs/mq/mqtt.rs [hunk 11]
```diff
@@ -104,16 +110,39 @@ impl L7ProtocolInfoInterface for MqttInfo {
...
+                    Some(t.clone())
+                } else {
+                    None
+                }
+            }
+            PacketKind::Unsubscribe | PacketKind::Subscribe => {
+                if let Some(s) = &self.subscribe_topics {
+                    let mut topic_str = String::new();
+                    for i in s {
+                        topic_str.push_str(format!("{},", i.name).as_str());
```
Comment (rvql) [inline agent/src/flow_generator/protocol_logs/mq/mqtt.rs:125]:
```suggestion
                        write!(&mut topic_str, "{},", i.name));
```

File: agent/src/flow_generator/protocol_logs/sql/redis.rs [hunk 12]
```diff
@@ -123,8 +131,25 @@ impl RedisInfo {
...
+        if other.is_on_blacklist {
+            self.is_on_blacklist = other.is_on_blacklist;
+        }
         Ok(())
     }
+
+    fn set_is_on_blacklist(&mut self, config: &LogParserConfig) {
+        if let Some(t) = config.l7_log_blacklist_trie.get(&L7Protocol::Redis) {
+            if t.request_resource.is_on_blacklist(unsafe {
+                // SAFETY: because it has been checked that every character is utf8
```
Comment (rvql) [inline agent/src/flow_generator/protocol_logs/sql/redis.rs:143]: 哪里检查的？

File: agent/src/config/handler.rs (hunk 1 above)
Comment (TomatoMr) [reply agent/src/config/handler.rs:710]: op 为空只是表示匹配还没有结束，所以还不能返回

File: agent/src/config/handler.rs (hunk 2 above)
Comment (TomatoMr) [reply agent/src/config/handler.rs]: 好的，已修改

File: agent/src/config/handler.rs (hunk 4 above)
Comment (TomatoMr) [reply agent/src/config/handler.rs]: 好的，已修改

File: agent/src/config/handler.rs (hunk 6 above)
Comment (TomatoMr) [reply agent/src/config/handler.rs]: 好的，已修改

File: agent/src/flow_generator/protocol_logs/mq/amqp.rs (hunk 8 above)
Comment (TomatoMr) [reply agent/src/flow_generator/protocol_logs/mq/amqp.rs]: 应该是之前写错了，已修改

File: agent/src/flow_generator/protocol_logs/mq/amqp.rs (hunk 9 above)
Comment (TomatoMr) [reply agent/src/flow_generator/protocol_logs/mq/amqp.rs]: 好的，已修改

File: agent/src/flow_generator/protocol_logs/mq/mqtt.rs (hunk 11 above)
Comment (TomatoMr) [reply agent/src/flow_generator/protocol_logs/mq/mqtt.rs]: 好的，已修改

File: agent/src/flow_generator/protocol_logs/sql/redis.rs (hunk 12 above)
Comment (TomatoMr) [reply agent/src/flow_generator/protocol_logs/sql/redis.rs]: 我以为是在 CommandLine::new 的时候检查了，已修改

**Review by rvql — APPROVED** [review]

**Review by rvql — APPROVED** [review]

**Review by rvql — APPROVED** [review]

**Pull request merged**